src/main/scala/com/mpc/scalats/core/ScalaParser.scala

```diff
@@ -4,24 +4,139 @@ package com.mpc.scalats.core
-    sealedTraitTypes ++
-      caseClassTypes.flatMap(getInvolvedTypes(Set.empty)).filter(isCaseClass).distinct.map(parseCaseClass(_, sealedTraitTypes)).distinct
+  def parseTypes(userInputTypes: List[Type]): List[CaseClass] = {
+
+    val initialTypes = userInputTypes.map(t => TypeToParse(t, None))
+
+    val allInvolvedTypes = initialTypes.flatMap(analyseType).distinct
+
+    allInvolvedTypes.map(parse)
+
   }
 
-  private def isSealedTrait(scalaType: Type) = {
+  private def isSealedTrait(scalaType: TypeToParse) = {
     scalaType.typeSymbol.asClass.isSealed &&
       scalaType.typeSymbol.asClass.isTrait
   }
 
-  private def parseSealedTrait(sealedTraitType: Type): CaseClass = {
+  private def analyseType(analysedType: TypeToParse): List[TypeToParse] = {
+    if (!alreadyExamined.exists(x => x.t == analysedType.t)) {
+      alreadyExamined.add(analysedType)
+      if (isSealedTrait(analysedType)) {
+        analysedType :: (directKnownSubclasses(analysedType.t).map(
+          t => TypeToParse(t, Some(analysedType.t))
+        ).flatMap(analyseType))
+
+      } else if (isCaseClass(analysedType)) {
+        analysedType :: getInvolvedTypes(analysedType).flatMap(analyseType)
+      } else {
+        Nil
+      }
+    } else {
+      Nil
+    }
+  }
+
+  private def directKnownSubclasses(tpe: Type): List[Type] = {
+    // Workaround for SI-7046: https://issues.scala-lang.org/browse/SI-7046
+    val tpeSym = tpe.typeSymbol.asClass
+
+    @annotation.tailrec
+    def allSubclasses(path: Traversable[Symbol], subclasses: Set[Type]): Set[Type] = path.headOption match {
+      case Some(cls: ClassSymbol) if (
+        tpeSym != cls && cls.selfType.baseClasses.contains(tpeSym)) => {
+        val newSub: Set[Type] = if (!cls.isCaseClass) {
+          logger.warn(s"cannot handle class ${cls.fullName}: no case accessor")
+          Set.empty
+        } else if (cls.typeParams.nonEmpty) {
+          logger.warn(s"cannot handle class ${cls.fullName}: type parameter not supported")
+          Set.empty
+        } else Set(cls.selfType)
+
+        allSubclasses(path.tail, subclasses ++ newSub)
+      }
+
+      case Some(o: ModuleSymbol) if (
+        o.companion == NoSymbol && // not a companion object
+          o.typeSignature.baseClasses.contains(tpeSym)) =>
+        allSubclasses(path.tail, subclasses + o.typeSignature)
+
+      case Some(o: ModuleSymbol) if (
+        o.companion == NoSymbol // not a companion object
+        ) => allSubclasses(path.tail, subclasses)
+
+      case Some(_) => allSubclasses(path.tail, subclasses)
+
+      case _ => subclasses
+    }
+
+    if (tpeSym.isSealed && tpeSym.isAbstract) {
+      allSubclasses(tpeSym.owner.typeSignature.decls, Set.empty).toList
+    } else List.empty
+  }
+
+  private def getInvolvedTypes(typeToParse: TypeToParse): List[TypeToParse] = {
+
+    var alreadyExamined : ArrayBuffer[Type] = ArrayBuffer.empty
+
+    def getInvolvedTypes(alreadyExamined: Set[Type])(scalaType: Type): List[Type] = {
```

Comment:
I do not understand what this function does ...

Reply:
It finds all "involved" types. That is : 

- types that are type arguments of the class
- types that are types of case class members
- types that are types arguments of such two things (recurisvely).

So for example if you have a member or type arg that is an Option[Class1[Class2]] it will detect Option, Class1, and Class2 through the recursion.

I'll add this explanation as a comment :-D